pkg/resource/deploy/step_generator.go

```diff
@@ -277,13 +353,22 @@ func (sg *stepGenerator) GenerateReadSteps(event ReadResourceEvent) ([]Step, err
 //
 // If the given resource is a custom resource, the step generator will invoke Diff and Check on the
 // provider associated with that resource. If those fail, an error is returned.
-func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, error) {
-	steps, err := sg.generateSteps(event)
+func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, bool, error) {
+	steps, async, err := sg.generateSteps(event)
 	if err != nil {
 		contract.Assertf(len(steps) == 0, "expected no steps if there is an error")
-		return nil, err
+		contract.Assertf(!async, "expected no async marker if there is an error")
+		return nil, false, err
+	}
+	if async {
```

Comment:
Might be worth a comment here about why we early exit/what it means to validate steps and when that should happen etc.?